iterator_test.go

```diff
@@ -2737,3 +2737,176 @@ func BenchmarkSeekPrefixTombstones(b *testing.B) {
 		iter.SeekPrefixGE(seekKey)
 	}
 }
+
+func runBenchmarkQueueWorkload(b *testing.B, deleteRatio float32, initOps int, valueSize int) {
+	const queueCount = 8
+	// These should be large enough to assign a unique key to each item in the
+	// queue.
+	const maxQueueIDLen = 1
+	const maxItemLen = 7
+	const maxKeyLen = maxQueueIDLen + 1 + maxItemLen
+	queueIDKeyspace := testkeys.Alpha(maxQueueIDLen)
+	itemKeyspace := testkeys.Alpha(maxItemLen)
+	key := make([]byte, maxKeyLen+1+len(fmt.Sprint(b.N)))
+	val := make([]byte, valueSize)
+	rng := rand.New(rand.NewSource(0))
+
+	getKey := func(q int, i int) []byte {
+		n := testkeys.WriteKey(key, queueIDKeyspace, int64(q))
+		key[n] = '/'
+		prefixLen := n + 1
+		n = testkeys.WriteKey(key[prefixLen:], itemKeyspace, int64(i))
+		return key[:prefixLen+n]
+	}
+
+	type Queue struct {
+		start int
+		end   int // exclusive
+	}
+	var queues = make([]*Queue, queueCount)
+	for i := 0; i < queueCount; i++ {
+		queues[i] = &Queue{}
+	}
+
+	o := (&Options{
+		DisableWAL:         true,
+		FS:                 vfs.NewMem(),
+		Comparer:           testkeys.Comparer,
+		FormatMajorVersion: FormatNewest,
+	}).EnsureDefaults()
+
+	d, err := Open("", o)
+	require.NoError(b, err)
+
+	processQueueOnce := func(batch *Batch, batchSize int) *Batch {
+		for {
+			// Randomly pick a queue to process.
+			q := rng.Intn(queueCount)
+			queue := queues[q]
+
+			isDelete := rng.Float32() < deleteRatio
+
+			if isDelete {
+				// Only process the queue if it's not empty. Otherwise, retry
+				// with a different queue.
+				if queue.start < queue.end {
+					require.NoError(b, batch.Delete(getKey(q, queue.start), nil))
+					queue.start++
+					break
+				}
+			} else {
+				// Append to the queue.
+				require.NoError(b, batch.Set(getKey(q, queue.end), val, nil))
+				queue.end++
+				break
+			}
+		}
+
+		if batch.Len() >= batchSize {
+			require.NoError(b, batch.Commit(NoSync))
+			d.AsyncFlush()
+			return d.NewBatch()
+		}
+		return batch
+	}
+
+	// First, process queues initialOps times.
+	batch := d.NewBatch()
+	for i := 0; i < initOps; i++ {
+		// Use a large batch size to speed up initialization.
+		batch = processQueueOnce(batch, 10<<24 /* 167 MiB */)
+	}
+	require.NoError(b, batch.Commit(NoSync))
+
+	b.Run(fmt.Sprintf("initial_ops=%d,deleteRatio=%.2f,valueSize=%d", initOps, deleteRatio, valueSize), func(b *testing.B) {
```

Comment:
This should be in the outer function, ie. `BenchmarkQueueWorkload`:

```
for _, deleteRatio := range deleteRatios {
		for _, valueSize := range valueSizes {
			for _, numInitOps := range initOps {
			  b.Run(fmt.Sprintf("initial_ops=%d/deleteRatio=%.2f/valueSize=%d", initOps, deleteRatio, valueSize), func(b *testing.B) {
			  runBenchmarkQueueWorkload(...)
			 })
    }
   }
}
```

It might sound like a subtle difference but currently we'd be doing the initialization for all the combinations of the benchmark even if we're only filtering and running one of them.

Reply:
Good catch, I've fixed this